File: packages/audit-trail/python/src/audit_trail/storage/file.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 MuVeraAI Corporation

"""
Append-only file storage backend.

Records are stored one JSON object per line (NDJSON / JSON Lines format).
The file is opened in append mode on construction and never truncated or
rewritten — callers relying on immutability should secure the file with
OS-level permissions (e.g., chown root + chmod 444 after rotation).

Reading always parses the entire file from disk so that the in-process view
stays consistent with anything written by concurrent processes.
"""

from __future__ import annotations

import json
import os
from pathlib import Path

import aiofiles
import aiofiles.os

from audit_trail.storage.interface import AuditStorage
from audit_trail.types import AuditFilter, AuditRecord


def _apply_filter(records: list[AuditRecord], audit_filter: AuditFilter) -> list[AuditRecord]:
    results: list[AuditRecord] = list(records)

    if audit_filter.agent_id is not None:
        agent_id = audit_filter.agent_id
        results = [r for r in results if r.agent_id == agent_id]

    if audit_filter.action is not None:
        action = audit_filter.action
        results = [r for r in results if r.action == action]

    if audit_filter.permitted is not None:
        permitted = audit_filter.permitted
        results = [r for r in results if r.permitted == permitted]

    if audit_filter.start_time is not None:
        start_time = audit_filter.start_time
        results = [r for r in results if r.timestamp >= start_time]

    if audit_filter.end_time is not None:
        end_time = audit_filter.end_time
        results = [r for r in results if r.timestamp <= end_time]

    offset = audit_filter.offset or 0
    results = results[offset:]

    if audit_filter.limit is not None:
        results = results[: audit_filter.limit]

    return results


class FileStorage(AuditStorage):
    """
    Persistent, append-only NDJSON file storage backend.

    Parameters
    ----------
    file_path:
        Path to the NDJSON file.  The file is created if it does not exist.
    """

    def __init__(self, file_path: str | Path) -> None:
        self._file_path = Path(file_path)

    async def append(self, record: AuditRecord) -> None:
        # model_dump with mode="json" ensures all values are JSON-serialisable.
        line = json.dumps(record.model_dump(mode="json", exclude_none=False)) + "\n"
        async with aiofiles.open(self._file_path, mode="a", encoding="utf-8") as file_handle:
            await file_handle.write(line)

    async def query(self, audit_filter: AuditFilter) -> list[AuditRecord]:
        all_records = await self.all()
        return _apply_filter(all_records, audit_filter)

    async def all(self) -> list[AuditRecord]:
        if not self._file_path.exists():
            return []

        records: list[AuditRecord] = []
        async with aiofiles.open(self._file_path, mode="r", encoding="utf-8") as file_handle:
            async for line in file_handle:
                stripped = line.strip()
                if not stripped:
                    continue
                try:
                    data = json.loads(stripped)
                    records.append(AuditRecord.model_validate(data))
                except (json.JSONDecodeError, Exception):
                    # Malformed lines are skipped — the hash chain verifier
                    # will detect any gaps caused by corruption.
                    pass

        return records

    async def count(self) -> int:
        all_records = await self.all()
        return len(all_records)

    @staticmethod
    def read_last_line_sync(file_path: str | Path) -> str | None:
        """
        Read the last non-empty line of the file synchronously.

        Used during construction to restore the chain's last hash without
        loading the full record set into memory.
        """
        path = Path(file_path)
        if not path.exists():
            return None
        with open(path, encoding="utf-8") as file_handle:
            lines = [line.strip() for line in file_handle if line.strip()]
        return lines[-1] if lines else None

File: packages/audit-trail/typescript/src/storage/file.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 MuVeraAI Corporation

import { createWriteStream, existsSync, readFileSync, WriteStream } from "node:fs";
import { readFile } from "node:fs/promises";
import type { AuditFilter, AuditRecord } from "../types.js";
import type { AuditStorage } from "./interface.js";

/**
 * Append-only file storage backend.
 *
 * Records are stored one JSON object per line (NDJSON / JSON Lines format).
 * The file is opened in append mode on construction and never truncated or
 * rewritten — callers relying on immutability should secure the file with
 * OS-level permissions.
 *
 * Reading always parses the entire file from disk so that the in-process
 * view stays consistent with anything written by concurrent processes.
 */
export class FileStorage implements AuditStorage {
  private readonly filePath: string;
  private readonly writeStream: WriteStream;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.writeStream = createWriteStream(filePath, { flags: "a", encoding: "utf8" });
  }

  async append(record: AuditRecord): Promise<void> {
    const line = JSON.stringify(record) + "\n";
    await new Promise<void>((resolve, reject) => {
      this.writeStream.write(line, (error) => {
        if (error !== null && error !== undefined) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
  }

  async query(filter: AuditFilter): Promise<AuditRecord[]> {
    const allRecords = await this.all();
    return applyFilter(allRecords, filter);
  }

  async all(): Promise<AuditRecord[]> {
    if (!existsSync(this.filePath)) {
      return [];
    }

    const content = await readFile(this.filePath, "utf8");
    const lines = content
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line.length > 0);

    const records: AuditRecord[] = [];
    for (const line of lines) {
      try {
        const parsed = JSON.parse(line) as AuditRecord;
        records.push(parsed);
      } catch {
        // Malformed lines are skipped — the hash chain verifier will catch
        // any gaps caused by corruption.
      }
    }

    return records;
  }

  async count(): Promise<number> {
    const records = await this.all();
    return records.length;
  }

  /**
   * Close the underlying write stream. Call this when the storage is no
   * longer needed to release the file handle.
   */
  close(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      this.writeStream.end((error: Error | null | undefined) => {
        if (error !== null && error !== undefined) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
  }

  /**
   * Read the last line of the file synchronously. Used during construction
   * to restore the chain's last hash without loading the full record set.
   */
  static readLastLineSynchronously(filePath: string): string | null {
    if (!existsSync(filePath)) {
      return null;
    }
    const content = readFileSync(filePath, "utf8");
    const lines = content
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line.length > 0);
    return lines.length > 0 ? (lines[lines.length - 1] ?? null) : null;
  }
}

function applyFilter(records: AuditRecord[], filter: AuditFilter): AuditRecord[] {
  let results = records;

  if (filter.agentId !== undefined) {
    const agentId = filter.agentId;
    results = results.filter((record) => record.agentId === agentId);
  }

  if (filter.action !== undefined) {
    const action = filter.action;
    results = results.filter((record) => record.action === action);
  }

  if (filter.permitted !== undefined) {
    const permitted = filter.permitted;
    results = results.filter((record) => record.permitted === permitted);
  }

  if (filter.startTime !== undefined) {
    const startTime = filter.startTime;
    results = results.filter((record) => record.timestamp >= startTime);
  }

  if (filter.endTime !== undefined) {
    const endTime = filter.endTime;
    results = results.filter((record) => record.timestamp <= endTime);
  }

  const offset = filter.offset ?? 0;
  results = results.slice(offset);

  if (filter.limit !== undefined) {
    results = results.slice(0, filter.limit);
  }

  return results;
}
